sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala

```diff
@@ -618,7 +617,7 @@ private[hive] trait HiveInspectors {
         case x: DateObjectInspector if x.preferWritable() =>
           data: Any => {
             if (data != null) {
-              DateTimeUtils.fromJavaDate(x.getPrimitiveWritableObject(data).get())
+              x.getPrimitiveWritableObject(data).getDays
```

Comment:
`DateWritable` stores days since epoch. Here need to just restore the value.

Reply:
how about dates before 1582?

Reply:
For consistency, we should rebase them but round trip conversions:
```scala
hiveIo.DateWritable(DateTimeUtils.toJavaDate(value.asInstanceOf[Int]))
```
```scala
DateTimeUtils.fromJavaDate(x.getPrimitiveWritableObject(data).get())
```

The bug is either in Spark's toJavaDate/fromJavaDate or in Hive's DateWritable. Hive does some heavy/dangerous conversions https://github.com/apache/hive/blob/beb57c6e43586ec8be770975a009b955c8a9840e/storage-api/src/java/org/apache/hadoop/hive/serde2/io/DateWritable.java#L144-L184 . I would prefer to avoid it.

Reply:
I rebased days